File: java_solutions/3116.kth-smallest-amount-with-single-denomination-combination.java
/*
 * @lc app=leetcode id=3116 lang=java
 *
 * [3116] Kth Smallest Amount With Single Denomination Combination
 */

class Solution { public long findKthSmallest(int[] coins, int k) { return ((Function<Object[], Long>) gcdBox -> ((Function<LongBinaryOperator, Long>) gcd -> ((Function<LongBinaryOperator, Long>) lcm -> ((Function<long[], Long>) lcms -> ((Function<Object[], Long>) bsBox -> ((Function<Function<long[], Long>, Long>) bs -> bs.apply(new long[]{1L, Long.MAX_VALUE})).apply((Function<long[], Long>) (bsBox[0] = (Function<long[], Long>) state -> state[0] >= state[1] ? state[0] : ((LongUnaryOperator) mid -> Arrays.stream(lcms).map(l -> mid / l).sum() < k ? ((Function<long[], Long>) bsBox[0]).apply(new long[]{mid + 1L, state[1]}) : ((Function<long[], Long>) bsBox[0]).apply(new long[]{state[0], mid})).applyAsLong(state[0] + (state[1] - state[0]) / 2L)))).apply(new Object[1])).apply(IntStream.range(1, 1 << coins.length).mapToLong(i -> IntStream.range(0, coins.length).filter(j -> (i & (1 << j)) > 0).mapToLong(j -> (long) coins[j]).reduce(1L, lcm) * (Integer.bitCount(i) % 2 == 1 ? 1L : -1L)).toArray())).apply((a, b) -> a * (b / gcd.applyAsLong(a, b)))).apply((LongBinaryOperator) (gcdBox[0] = (LongBinaryOperator) (a, b) -> b == 0L ? a : ((LongBinaryOperator) gcdBox[0]).applyAsLong(b, a % b)))).apply(new Object[1]); } }

File: kotlin_solutions_1/3116.kth-smallest-amount-with-single-denomination-combination.kt
/*
 * @lc app=leetcode id=3116 lang=kotlin
 *
 * [3116] Kth Smallest Amount With Single Denomination Combination
 */

class Solution {
    fun findKthSmallest(coins: IntArray, k: Int): Long = DeepRecursiveFunction<Pair<Long, Long>, Long> { (a, b) -> if (b == 0L) a else callRecursive(b to (a % b)) }.let { gcd -> { a: Long, b: Long -> a / gcd(a to b) * b }.let { lcm -> (1 until (1 shl coins.size)).map { mask -> coins.indices.filter { (mask and (1 shl it)) != 0 }.fold(1L) { acc, i -> lcm(acc, coins[i].toLong()) } to (if (Integer.bitCount(mask) % 2 == 1) 1L else -1L) }.let { subsets -> { mid: Long -> subsets.sumOf { (l, sign) -> (mid / l) * sign } }.let { count -> DeepRecursiveFunction<Pair<Long, Long>, Long> { (low, high) -> if (low >= high) low else ((low + high) / 2).let { mid -> if (count(mid) >= k) callRecursive(low to mid) else callRecursive((mid + 1) to high) } }.invoke(1L to 50000000000000000L) } } } }
}